internal/postgres/reconcile.go

```diff
@@ -258,3 +261,146 @@ func PodSecurityContext(cluster *v1beta1.PostgresCluster) *corev1.PodSecurityCon
+	inDataVolume, inWALVolume *corev1.PersistentVolumeClaim,
+) (batchv1.Job, error) {
+
+	// create the pg_upgrade Job
+	upgradeJob := &batchv1.Job{}
+	upgradeJob.ObjectMeta = naming.PGUpgradeJob(cluster)
+
+	// set labels and annotations
+	var labels, annotations map[string]string
+	labels = naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetLabelsOrNil(),
+		naming.PGUpgradeJobLabels(cluster.Name))
+	annotations = naming.Merge(cluster.Spec.Metadata.GetAnnotationsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetAnnotationsOrNil(),
+	)
+	upgradeJob.ObjectMeta.Labels = labels
+	upgradeJob.ObjectMeta.Annotations = annotations
+
+	certVolumeMount := corev1.VolumeMount{
+		Name:      naming.CertVolume,
+		MountPath: naming.CertMountPath,
+		ReadOnly:  true,
+	}
+	certVolume := corev1.Volume{
+		Name: certVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				// PostgreSQL expects client certificate keys to not be readable
+				// by any other user.
+				// - https://www.postgresql.org/docs/current/libpq-ssl.html
+				DefaultMode: initialize.Int32(0o600),
+				Sources: []corev1.VolumeProjection{
+					{Secret: inClusterCertificates},
+					{Secret: inClientCertificates},
+				},
+			},
+		},
+	}
+
+	dataVolumeMount := DataVolumeMount()
+	dataVolume := corev1.Volume{
+		Name: dataVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: inDataVolume.Name,
+				ReadOnly:  false,
+			},
+		},
+	}
+	//while true; do sleep 30; done;
+	script := fmt.Sprintf(`OLDVERSION=%d
+NEWVERSION=%d
+echo -e "Performing PostgreSQL upgrade from version $OLDVERSION to $NEWVERSION for cluster %s.\n"
+cd /pgdata
+
+echo -e "Step 1: Making new pgdata directory...\n"
+mkdir /pgdata/pg$NEWVERSION
+
+echo -e "Step 2: Initializing new pgdata directory...\n"
+/usr/pgsql-$NEWVERSION/bin/initdb -k -D /pgdata/pg$NEWVERSION
+
+echo -e "\nStep 3: Setting the expected permissions on the old pgdata directory...\n"
+chmod 700 /pgdata/pg$OLDVERSION
+
+echo -e "Step 4: Copying shared_preload_libraries setting to new postgresql.conf file...\n"
+grep shared_preload_libraries /pgdata/pg$OLDVERSION/postgresql.conf >> /pgdata/pg$NEWVERSION/postgresql.conf
+
+echo -e "Step 5: Running pg_upgrade check...\n"
+time /usr/pgsql-$NEWVERSION/bin/pg_upgrade --old-bindir /usr/pgsql-$OLDVERSION/bin --new-bindir \
+/usr/pgsql-$NEWVERSION/bin --old-datadir /pgdata/pg$OLDVERSION --new-datadir /pgdata/pg$NEWVERSION --link --check
+
+echo -e "\nStep 6: Running pg_upgrade...\n"
+time /usr/pgsql-$NEWVERSION/bin/pg_upgrade --old-bindir /usr/pgsql-$OLDVERSION/bin --new-bindir \
+/usr/pgsql-$NEWVERSION/bin --old-datadir /pgdata/pg$OLDVERSION --new-datadir /pgdata/pg$NEWVERSION --link
+
+echo -e "\nStep 7: Copying patroni.dynamic.json...\n"
+cp /pgdata/pg$OLDVERSION/patroni.dynamic.json /pgdata/pg$NEWVERSION
```

Comment:
:pick: Some links to the Patroni docs would be good around here to show where steps such as this one (deleting the `patroni.dynamic.json`) come from.

Reply:
:+1: I added a comment and a link for this step